scripts/tests/run_python_test.py

```diff
@@ -239,36 +299,95 @@ def main_impl(app: str, factory_reset: bool, factory_reset_app_only: bool, app_a
-                                     f_stderr=stream_output)
+                                     f_stderr=sys.stderr.buffer)
     test_script_process.start()
     test_script_process.p.stdin.close()
-    test_script_exit_code = test_script_process.wait()
-
-    if test_script_exit_code != 0:
-        logging.error("Test script exited with returncode %d" % test_script_exit_code)
-
-    if app_process:
-        logging.info("Stopping app with SIGTERM")
-        if app_stdin_forwarding_thread:
-            app_stdin_forwarding_stop_event.set()
-            app_stdin_forwarding_thread.join()
-        app_process.terminate()
-        app_exit_code = app_process.returncode
-
-    # We expect both app and test script should exit with 0
-    exit_code = test_script_exit_code or app_exit_code
-
-    if quiet:
-        if exit_code:
-            sys.stdout.write(stream_output.getvalue().decode('utf-8'))
-        else:
-            logging.info("Test completed successfully")
 
-    if exit_code != 0:
-        logging.error("SUBPROCESS failure: ")
-        logging.error("  TEST SCRIPT: %d (%r)", test_script_exit_code, final_script_command)
-        logging.error("  APP:         %d (%r)", app_exit_code, [app] + shlex.split(app_args))
-        sys.exit(exit_code)
+    try:
+        test_script_exit_code = test_script_process.wait()
+
+        if test_script_exit_code != 0:
+            logging.error("Test script exited with returncode %d" % test_script_exit_code)
+
+        # Stop the restart monitor thread if it exists
+        if reboot_dut:
+            if restart_monitor_thread and restart_monitor_thread.is_alive():
+                logging.info("Stopping app restart monitor thread")
+                restart_monitor_thread.join(2.0)
+
+        # Get the current app manager if it exists
+        current_app_manager = None
+        if app_manager_ref:
+            with app_manager_lock:
+                current_app_manager = app_manager_ref[0]
+
+        if current_app_manager:
+            logging.info("Stopping app with SIGTERM")
+            current_app_manager.stop()
+            if current_app_manager.get_process():
+                app_exit_code = current_app_manager.get_process().returncode
+
+        # We expect both app and test script should exit with 0
+        exit_code = test_script_exit_code or app_exit_code
+
+        if quiet:
+            if exit_code:
+                sys.stdout.write(stream_output.getvalue().decode('utf-8'))
+            else:
+                logging.info("Test completed successfully")
+
+        if exit_code != 0:
+            logging.error("SUBPROCESS failure: ")
+            logging.error("  TEST SCRIPT: %d (%r)", test_script_exit_code, final_script_command)
+            logging.error("  APP:         %d (%r)", app_exit_code, [app] + shlex.split(app_args))
+            sys.exit(exit_code)
+
+    finally:
+        # Stop the restart monitor thread if it exists
+        if reboot_dut:
```

Comment:
this was done earlier - is there an expectation that the earlier try loop will fail before that happens?